src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -188,39 +234,61 @@ public async Task UpdatePrimaryBranchSolutionAsync(AssetProvider assetProvider,
                 }
 
-                var solution = await CreateSolution_NoLockAsync(
-                    assetProvider,
-                    solutionChecksum,
-                    fromPrimaryBranch,
-                    workspaceVersion,
-                    CurrentSolution,
-                    cancellationToken).ConfigureAwait(false);
+                return GetProjectSubsetSolutionSlowAsync(box?.Value.solution ?? CurrentSolution, assetProvider, solutionChecksum, projectId, cancellationToken);
 
-                _lastRequestedSolutionWithChecksum = Tuple.Create(solutionChecksum, solution);
+                async ValueTask<Solution> GetProjectSubsetSolutionSlowAsync(
+                    Solution baseSolution,
+                    AssetProvider assetProvider,
+                    Checksum solutionChecksum,
+                    ProjectId projectId,
+                    CancellationToken cancellationToken)
+                {
+                    var updater = new SolutionCreator(Services.HostServices, assetProvider, baseSolution, cancellationToken);
 
-                return solution;
+                    // check whether solution is update to the given base solution
+                    Solution result;
+                    if (await updater.IsIncrementalUpdateAsync(solutionChecksum).ConfigureAwait(false))
+                    {
+                        // create updated solution off the baseSolution
+                        result = await updater.CreateSolutionAsync(solutionChecksum).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        // we need new solution. bulk sync all asset for the solution first.
+                        await assetProvider.SynchronizeSolutionAssetsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                        // get new solution info and options
+                        var (solutionInfo, options) = await assetProvider.CreateSolutionInfoAndOptionsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                        var workspace = new TemporaryWorkspace(Services.HostServices, WorkspaceKind.RemoteTemporaryWorkspace, solutionInfo, options);
+                        result = workspace.CurrentSolution;
+                    }
+
+                    // Cache the result of our computation.  Note: this is simply a last caller wins strategy.  However,
+                    // in general this should be fine as we're primarily storing this to make future calls to synchronize
+                    // this project cone fast.
+                    _lastRequestedProjectIdToSolutionWithChecksum[projectId] = new((solutionChecksum, result));
+                    return result;
+                }
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken))
+            {
+                throw ExceptionUtilities.Unreachable;
             }
         }
```

Comment:
Did you mean to put this filter inside the locla function. The outer method doesn't have await, so this doesn't really do anything useful.

Reply:
will fixup.